core/app/[locale]/(default)/product/[slug]/_actions/wishlist-action.ts

```diff
@@ -0,0 +1,301 @@
+        }
+      }
+    }
+  }
+`);
+
+const AddToWishlistMutation = graphql(`
+  mutation AddToWishlistMutation($wishlistId: Int!, $productId: Int!, $variantId: Int) {
+    wishlist {
+      addWishlistItems(
+        input: {
+          entityId: $wishlistId
+          items: [{ productEntityId: $productId, variantEntityId: $variantId }]
+        }
+      ) {
+        result {
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+const DeleteWishlistItemsMutation = graphql(`
+  mutation DeleteWishlistItemsMutation($wishlistId: Int!, $wishlistItemId: Int!) {
+    wishlist {
+      deleteWishlistItems(input: { entityId: $wishlistId, itemEntityIds: [$wishlistItemId] }) {
+        result {
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+const menuItemSchema = z.object({
+  action: z.enum(['add', 'remove']),
+  wishlistId: z.number().nonnegative(),
+  wishlistItemId: z.number().optional(),
+  redirectTo: z.string(),
+});
+
+const menuItemParser = z.string().transform((str, ctx) => {
+  try {
+    return menuItemSchema.parse(JSON.parse(str));
+  } catch {
+    ctx.addIssue({ code: 'custom', message: 'Invalid menuItem payload.' });
+
+    return z.NEVER;
+  }
+});
+
+const schema = z.object({
+  productId: z.number().nonnegative().min(1),
+  selectedSku: z.string(),
+  menuItem: menuItemParser,
+});
+
+interface WishlistAddMutationVariables {
+  wishlistId: number;
+  productId: number;
+  variantId?: number;
+}
+
+interface WishlistRemoveMutationVariables {
+  wishlistId: number;
+  wishlistItemId: number;
+}
+
+const getVariantIdFromSku = cache(
+  async (productId: number, sku: string, customerAccessToken?: string) => {
+    const { data } = await client.fetch({
+      document: VariantIdFromSkuQuery,
+      variables: { productId, sku },
+      customerAccessToken,
+      fetchOptions: { next: { revalidate } },
+    });
+
+    return data.site.product?.variants.edges?.[0]?.node.entityId ?? undefined;
+  },
+);
+
+async function addToDefaultWishlist(
+  customerAccessToken: string,
+  wishlistName: string,
+  productId: number,
+  variantId?: number,
+) {
+  const { data } = await client.fetch({
+    document: CreateWishlistMutation,
+    variables: {
+      input: {
+        name: wishlistName,
+        isPublic: false,
+        items: [{ productEntityId: productId, variantEntityId: variantId }],
+      },
+    },
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.createWishlist?.result) {
+    throw new Error('Failed to add item to default wishlist. Response was empty.');
+  }
+}
+
+async function addToWishlist(customerAccessToken: string, variables: WishlistAddMutationVariables) {
+  const { data } = await client.fetch({
+    document: AddToWishlistMutation,
+    variables,
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.addWishlistItems?.result) {
+    throw new Error('Failed to add item to wishlist. Response was empty.');
```

Comment:
🍹 Something to think about here, if we are throwing a bunch of generic errors with custom message it might make sense to create our own custom errors. Even if it seems like a lot right now, it allows consumers to extend functionality if they ever have the need, e.g. `error instanceof CustomError`. It helps not checking `error.message.includes('some text')` to see what the error type is.